### ISSUE TITLE
sth still confuses about nnPUloss

### ISSUE THREAD

**Issue by Wongcheukwai** [VITA-Group/Self-PU#4]
Thank you for pointing out this repo for me: https://github.com/kiryor/nnPUlearning/blob/a522c0edd22c54b627eb131e960584950eae5330/pu_loss.py#L51.

However, this is the part about the LOSS:

        if self.nnpu:
            if negative_risk.data < -self.beta:
                objective = positive_risk - self.beta
                self.x_out = -self.gamma * negative_risk
            else:
                self.x_out = objective
        else:
            self.x_out = objective
        self.loss = xp.array(objective.data, dtype=self.x_out.data.dtype)
        return self.loss,

as you can see, if negative_risk.data < -self.beta, objective is assigned with positive_risk(instead of negative loss) and the self.loss is equal to objective.data (who data.dtype is self.x_out, which is the datatype of negetive_risk).
 So I guess you have to return postive_loss here?

**Comment by chanyh0** [comment]
The key is self.x_out, which is used in backward function. So the actual optimization objective (for calculating the gradient) is self.x_out, which is -self.gamma * negative _risk.

**Comment by Wongcheukwai** [comment]
oh yes TAT, thank you kindly for your patience.

**Comment by chanyh0** [comment]
You are welcome :)
